### PR TITLE
When "assigned", a `discard` doesn't escape out of the current expression.

### LINKED ISSUE #39663: Compiler should allow assignment to discard - from ref-struct returning method.
**Version Used**: 
C# 8 .NET Core 3.0 (original)
C# 9 .NET 5 (updated)

```
    public void Foo()
    {
        var dummy = Bar(items: stackalloc int[5]); // OK, assigning return value to variable.
        
        Bar(items: stackalloc int[5]); // OK, ignoring return value.
        
        _ = Bar(items: stackalloc int[5]); // Error on assignment of return value to discard.
    }

    ReadOnlySpan<int> Bar(ReadOnlySpan<int> items) => items;
```

**Expected Behavior**:
The compiler should allow the assignment to discard because it allows the assignment to a variable. 

**Actual Behavior**:
The compiler complains about exposing the span outside its containing method.

(06/07/2021) edit: The code was updated to include the case where the return value is ignored. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAOgCUBXGAFyjAFNCBhAezAAcoAbGgJwGVuA3KAGMaAZwDcmTAgDM+JLnq5MAb2UZcm/HIQAWXADFGjABQBKTFtwrLVrfwhdcAE1JgwAT1wBeXACFHEygKGjBREFxRCgghAGsIdnZGIVxYCgBtAFYAXTNxXAB6AtwAeQBpABpcCFFRKABzGFh6/AB2XAd2UhpcCkYOxygIYE5CWzt1Cc0AriCQsIiomPjE5NTKLNz8otLK1MbGLma2ga66cbsLqwB9H39A4NDwyOi4hKSUtM28wuKAUS4XEOuEYMGqtQaMFolBBADMTp1ur1+s4oKIhI5nGMNFoAL5SHGaYg0CDOEowdgeHgsCAwAA8aQAfPdZsTSeTKdTaQzKMzHmEzD4+fMJFZMLigA=)

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #39663.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jaredpar, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -2781,7 +2782,8 @@ private BoundExpression BindOutDeclarationArgument(DeclarationExpressionSyntax d
                         var declType = BindVariableTypeWithAnnotations(designation, diagnostics, typeSyntax, ref isConst, out isVar, out alias);
                         Debug.Assert(isVar != declType.HasType);
 
-                        return new BoundDiscardExpression(declarationExpression, declType.Type);
+                        // ValEscape is the same as for an uninitialized local
+                        return new BoundDiscardExpression(declarationExpression, Binder.ExternalScope, declType.Type);
                     }
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2787]: Why would this be external scope? Can you add a test like this:

```cs
Span<int> a = stackalloc int[10];
M(ref a, out _);

void M(ref Span<int> s, out Span<int> s1) => s1 = s;
```

**Comment by 333fred** [comment]
Done review pass (commit 1).

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2787]: Also, a symmetric test like

```cs
Span<int> a = new int[10];
M(ref a, out _);

void M(ref Span<int> s, out Span<int> s1) => s1 = s;
```

would also be good to have.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2787]: > Why would this be external scope? 

I think the added comment explains why. BTW, there is no change here for ```out``` discards. The scope used to be External and remains External.  Also, if the value is replaced with ```LocalScopeDepth``` the  Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.RefEscapingTests.DiscardExpressionSpan unit-test is failing with unexpected errors. So, I would say we are good with test coverage for this code path.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2787]: To be specific, the following scenarior is getting broken for example:
```
    static Span<int> Test1()
    {
        var s = ReturnsSpan(out var _);

        // ok
        return s;
    }
```

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2787]: Ah, so there's some negative downstream effects. This is interesting: we'd almost want `_` to not contribute to the scoping, either narrowing or expanding, though perhaps @jaredpar will point out some case that this wouldn't work in.

**Review by 333fred — APPROVED** [review]
I'll approve this, though it does feel like the test scenarios I pointed out should compile, at least their behavior isn't changing from the current behavior.

**Comment by AlekseyTs** [comment]
> it does feel like the test scenarios I pointed out should compile

I agree. The scenario with `stackalloc` doesn't compile today, similar to a scenario with an uninitialized local, or a local initialized with `new`. It looks like there is a more general flaw in analysis around `out` arguments, but I would say it is out of scope of this PR. Opened https://github.com/dotnet/roslyn/issues/56587 to track.

**Review by jaredpar — APPROVED** [review]

**Pull request merged**